design/keycloak.x/configuration.md

```diff
@@ -0,0 +1,159 @@
+would like to achieve something similar to https://quarkus.io/guides/all-config.
+
+In the documentation the property key used in the properties file should be used. It should include a description, the 
+type and the default value.
+
+We would also like the ability to list properties on the command-line by running `bin/kc.sh --list-config-options`.
+
+For SPIs and Providers we should extend the ProviderFactory to add an option method to return information on what
+properties it supports.
+
+## Config Options
+
+The aim is to start small with what config options we provide, and extend when there is a need. Great care should be 
+taken in reducing the amount of configuration needed, where opinionated/best-practice behaviour should be favoured, with
+high-level and simple configuration options introduced only when needed.
+
+Below is a list of config by category we should introduce initially:
+
+### HTTP
+
+Keycloak.X will not come with the capability to generate a self-signed certificate. Instead, Keycloak by default will
+complain if a certificate has not been configured, but will allow explicitly disabling SSL for development purposes.
+
+* http.mode - Options `ssl` (default), `unsecure`  
+* http.port
+* http.ssl-port
+* http.ssl.certificate.file
+* http.ssl.certificate.key-file
+* http.ssl.certificate.key-store-file
+* http.ssl.certificate.key-store-file-type
+* http.ssl.certificate.key-store-password
+* http.ssl.certificate.trust-store-file
+* http.ssl.certificate.trust-store-file-type
+* http.ssl.certificate.trust-store-password
+* http.ssl.cipher-suites
+* http.ssl.protocols
+* http.ssl.client-auth
+
+### Proxy Mode
+
+The proxy mode will allow an easy way to configure if Keycloak is hosted behind a reverse-proxy or not, including how
+SSL is terminated.
+
+* proxy.mode - Options are `none` (default), `reencrypt`, `edge` and `passthrough`.
+
+Setting `proxy.mode` will under the covers configure `http.mode` as well as `quarkus.http.proxy-address-forwarding`.
+
+## Database
+
+The aim is that a user should not need to understand Hibernate, JDBC drivers, etc. to be able to configure the database.
+As such we should introduce a high-level option `db.vendor` instead of requiring users to set JDBC driver and Hibernate
+dialects.
+
+We will support either `db.url` which will support a JDBC URL, or using separate properties to build-up the URL behind the
+scenes.
+
+* db.vendor
+* db.url 
+* db.hostname
+* db.port - defaults to the default for the configured vendor 
+* db.schema
+* db.database
+* db.username
+* db.password
+* db.pool.initialSize
+* db.pool.minSize - default to 0
+* db.pool.maxSize - default to 100
+
+## Caches (Infinispan and JGroups)
+
+As we aim to not use an embedded Infinispan server in the future, configuration of caches will not for now follow the
+standard way of configuring Keycloak. Instead there will be a separate XML file for configuring this (`conf/cluster.xml`).
```

Comment:
I'd prefer a uniform format for the configuration. Currently caches would be configured via xml, the others via .properties files.

Reply:
I appreciate that, but as stated the plan is to not use an embedded Infinispan server in the future, which would make cache configuration much simpler (would be closer to how you configure a DB with just specifying the url, not stacks, etc.).

We may be able to come up with some slightly nicer interim solution though.

Reply:
Did you see this? Referencing an jgroups config from infinispan config:
https://github.com/keycloak/keycloak-community/pull/154#issuecomment-645234340
Perhaps this could be the interim solution

Reply:
I think @pedroigor proposed something similar. I'd like to split this up though. First pass would be to just make caching configerable, then we can look into usability improvements around it later.

Reply:
Currently, users have complete control over the `cluster.xml` file so they can configure ISPN at their weill. Including doing something like that, @thomasdarimont.

However, I see as a good improvement to support defaults to make life easier for people running on well-known cloud providers or on-premise:

```
cluster.ec2.enabled
cluster.ec2.s3.region-name
```

```
cluster.kubernetes.enabled
cluster.kubernetes.dns.query
```

```
cluster.baremetal.tcp.hosts
cluster.baremetal.udp.multicast.address
```

```
# For Cross-DC on top of a particular stack something like
cluster.kubernetes.enabled
cluster.kubernetes.dns.query
cluster.cross-dc.enabled
cluster.cross-dc.option1
```

And so forth, where most of the configuration above would come from the defaults shipped with JGroups itself. Plus any other default that we make sense for running Keycloak on top of those platforms.

For more advanced use cases and for those looking to hack their own configuration, `cluster.xml`. 

@mposolda Can help us a lot here ...

Reply:
Closing this as I want to defer discussion around caches to a follow-up

Reply:
> However, I see as a good improvement to support defaults to make life easier for people running on well-known cloud providers or on-premise:
> 
> ```
> cluster.ec2.enabled
> cluster.ec2.s3.region-name
> ```
> 
> ```
> cluster.kubernetes.enabled
> cluster.kubernetes.dns.query
> ```

Probably not the best idea to provide the `cluster.*.enabled` options; thus, we will let the user "enable" multiple JGroups configs at once, which doesn't make sense. I'd rather suggest something like this:
```
cluster = ec2
# or maybe "cluster.stack = ec2", or "cluster.type = ec2"
cluster.ec2.foobar = ...
```

> 
> And so forth, where most of the configuration above would come from the defaults shipped with JGroups itself. Plus any other default that we make sense for running Keycloak on top of those platforms.

AFAIK, JGroups do not ship the defaults, but Infinispan [does](https://github.com/infinispan/infinispan/tree/master/core/src/main/resources/default-configs). It would be nice to have an ability to activate those defaults right from the Keycloak config.